contracts/FlowIDTableStaking.cdc

```diff
@@ -870,6 +870,79 @@ pub contract FlowIDTableStaking {
+            for nodeID in allNodeIDs {
+                let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
+                if nodeRecord.initialWeight == 100 {
+                    nodeCounts[nodeRecord.role] = nodeCounts[nodeRecord.role]! + 1
+                }
+            }
+
+            for key in nodeCounts.keys {
+                while nodeCounts[key]! > slotLimits[key]! {
+                    let selection = unsafeRandom() % UInt64(nodeCounts[key]!)
+                    var searchIndex: UInt64 = 0
+                    for nodeID in allNodeIDs {
+                        let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
+                        if nodeRecord.role == key && nodeRecord.initialWeight == 100 {
+                            if selection == searchIndex {
+                                emit NodeRemovedAndRefunded(nodeID: nodeRecord.id, amount: nodeRecord.tokensCommitted.balance + nodeRecord.tokensStaked.balance)
+
+                                // move their committed tokens back to their unstaked tokens
+                                nodeRecord.tokensUnstaked.deposit(from: <-nodeRecord.tokensCommitted.withdraw(amount: nodeRecord.tokensCommitted.balance))
+
+                                // Set their request to unstake equal to all their staked tokens
+                                // since they are forced to unstake
+                                nodeRecord.tokensRequestedToUnstake = nodeRecord.tokensStaked.balance
+
+                                // Iterate through all delegators and unstake their tokens
+                                // since their node has unstaked
+                                for delegator in nodeRecord.delegators.keys {
+                                    let delRecord = nodeRecord.borrowDelegatorRecord(delegator)
+
+                                    if delRecord.tokensCommitted.balance > 0.0 {
+                                        emit DelegatorTokensUnstaked(nodeID: nodeRecord.id, delegatorID: delegator, amount: delRecord.tokensCommitted.balance)
+
+                                        // move their committed tokens back to their unstaked tokens
+                                        delRecord.tokensUnstaked.deposit(from: <-delRecord.tokensCommitted.withdraw(amount: delRecord.tokensCommitted.balance))
+                                    }
+
+                                    // Request to unstake all tokens
+                                    delRecord.tokensRequestedToUnstake = delRecord.tokensStaked.balance
+                                }
+
+                                // Clear initial weight because the node is not staked any more
+                                nodeRecord.initialWeight = 0
+
+                                //Decrement the count for this role
+                                nodeCounts[key] = nodeCounts[key]! - 1
+
+                                break
```

Comment:
This is potentially quite expensive - in the worst case we will iterate `nodeIDs` once for each excess provisionally registered node. 
* ⚠️  I believe anyone can register a node, even with 0 stake - these nodes will stay in `nodeIDs` forever ([reference](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowIDTableStaking.cdc#L1190-L1199))
* Similarly, anyone can provisionally register a node by providing the stake minimum.

I would suggest taking a different approach to this iteration once this moves past the WIP stage.

Reply:
I'm refactoring this.